spylib/utils/session_token.py

```diff
@@ -0,0 +1,122 @@
+from __future__ import annotations
+
+from typing import Any, ClassVar, Dict, List, Optional
+from urllib.parse import urlparse
+
+from jwt import decode
+from pydantic.main import BaseModel
+from pydantic.networks import HttpUrl
+
+from .domain import store_domain
+
+
+class ValidationError(Exception):
+    pass
+
+
+class InvalidIssuerError(ValidationError):
+    pass
+
+
+class MismatchedHostError(ValidationError):
+    pass
+
+
+class TokenAuthenticationError(ValidationError):
+    pass
+
+
+class SessionToken(BaseModel):
+    """
+    Session tokens are derived from the authrization header from Shopify.
+    This performs the set of validations as defined by shopify
+    https://shopify.dev/apps/auth/session-tokens/authenticate-an-embedded-app-using-session-tokens#obtain-session-details-manually
+    """
+
+    iss: HttpUrl
+    dest: HttpUrl
+    aud: Optional[str]
+    sub: int
+    exp: Optional[float]
+    nbf: Optional[float]
+    iat: Optional[float]
+    jti: int
+    sid: str
+
+    algorithm: ClassVar[str] = "HS256"
+    prefix: ClassVar[str] = "Bearer "
+    required_fields: ClassVar[List[str]] = ["iss", "dest", "sub", "jti", "sid"]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def validate_self(self) -> None:
+        self.__validate_hostname()
+        self.__validate_destination()
+        self.__validate_sub()
+
+    @classmethod
+    def decode_token_from_header(
+        cls,
+        authorization_header: str,
+        api_key: str,
+        secret: str,
+    ) -> SessionToken:
+        # Verify the integrity of the token
+
+        # Take the authorization headers and unload them
+        token = cls.__extract_session_token(authorization_header)
+        payload = cls.__decode_session_token(token, api_key, secret)
+
+        # Verify enough fields specified and perform validation checks
+        session_token = cls(**payload)
```

Comment:
Read [this](https://github.com/orgs/SatelCreative/teams/python/discussions/43)
```suggestion
        session_token = cls.parse_obj(payload)
```